overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,121 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
+//
+// It is the caller's responsibility to lock state before calling this function.
+func snapRevisionsToKeep(st *state.State, snapName string) ([]snap.Revision, error) {
+	var snapst SnapState
+	err := Get(st, snapName, &snapst)
+	if err != nil {
+		return nil, err
+	}
+
+	var revisionsToKeep []snap.Revision
+
+	// keep revisions in snap's sequence
+	for _, seq := range snapst.Sequence {
+		revisionsToKeep = append(revisionsToKeep, seq.Revision)
+	}
+
+	// keep revision in refresh hint (if it exists)
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	} else if hint, ok := refreshHints[snapName]; ok {
+		revisionsToKeep = append(revisionsToKeep, hint.Revision())
+	}
+
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, _, err := snapSetupAndState(t)
+			if err != nil {
+				return nil, err
+			}
+			if snapsup.InstanceName() != snapName {
+				continue
+			}
+			revisionsToKeep = append(revisionsToKeep, snapsup.Revision())
+		}
+	}
+
+	return revisionsToKeep, nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
+	snapStates, err := All(st)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for snapName := range snapStates {
+		revisions, err := snapRevisionsToKeep(st, snapName)
+		if err != nil {
+			return err
+		}
+		for _, revision := range revisions {
+			mountFile := snap.MountFile(snapName, revision)
+			if osutil.FileExists(mountFile) {
+				keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirState(dirs.SnapBlobDir, "*.snap", keepContent)
+	return err
+}
+
+// cleanSnapDownloads removes downloads that are no longer needed for the given snap.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanSnapDownloads = func(st *state.State, snapName string) error {
+	revisions, err := snapRevisionsToKeep(st, snapName)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for _, revision := range revisions {
+		mountFile := snap.MountFile(snapName, revision)
+		if osutil.FileExists(mountFile) {
+			keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+		}
+	}
+
+	// XXX: could we need more digits?
+	const maxRevDigits = 10
+	var globs []string
+	for i := 1; i <= maxRevDigits; i++ {
+		var digitsPat string
+		for j := 1; j <= i; j++ {
+			digitsPat += "[0-9]"
+		}
```

Comment:
If we end up keeping this code, then we can simplify this part by using `strings.Repeat`
```suggestion
		digitsPat := strings.Repeat("[0-9]", i)
```

Reply:
this is much cleaner, thank you!

Reply:
my approach suffers from the unlikely but possible case that revisions are assigned large numbers > 10 digits. it says in the docs that currently it is incremented but that it is arbitrary https://snapcraft.io/docs/revisions.